Cell Centered Precursors

zachhardy opened the pull request:
## Notes

Currently, only the maximum number of precursors that exist on a material are stored per node. This is done to eliminate the possibility of potentially very sparse vectors in problems with many materials. With this infrastructure, nodally define precursors only make sense with PWLD and FV. With PWLC, nodally defined precursors do not make sense because interfaces between materials can have different precursor species on either side. The obvious issue here is that we have `max_precursors_per_material` slots to store up to `2 * max_precursors_per_material`. For this reason, it makes more sense to store cell averaged precursors. 

### Changes to `LinearBoltzmann::Solver`

- Added precursor initialization to `LinearBoltzmann::Solver::InitMaterials` and `LinearBoltzmann::Solver::InitPArrays`
- Added `LinearBoltzmann::Solver::ComputePrecursors` routine to compute steady state precursor concentrations.
- Call `ComputePrecursors` at the end of the `Execute` routine
- Added `num_groups`, `num_precursors`, and `max_precursors_per_material` attributes

zachhardy committed:
Added precursor stuff to LBSolver.

zachhardy committed:
Added precursor initializations.

zachhardy committed:
ComputePrecursors routine implemented.

zachhardy committed:
Compute precursors post-execution.

zachhardy committed:
Cleanup KEigenvalue::Solver.

zachhardy committed:
int -> size_t

zachhardy committed:
num_moments from int -> size_t

zachhardy committed:
Per Jan's request.

Naktakala commented on ChiModules/LinearBoltzmannSolver/lbs_01b_initmaterials.cc:
```diff
@@ -118,11 +118,26 @@ void LinearBoltzmann::Solver::InitMaterials(std::set<int>& material_ids)
       << material_xs[matid_to_xs_map[mat_id]]->transfer_matrices.size() << "\n";
   }//for material id
 
+  num_groups = int(groups.size());
```
This `int(groups.size())` cast should not be necessary and its c-style. Just remove it please.

zachhardy replied:
Done.

Naktakala commented on ChiModules/LinearBoltzmannSolver/lbs_compute_precursors.cc:
```diff
@@ -0,0 +1,61 @@
+void LinearBoltzmann::Solver::ComputePrecursors()
+{
+  auto pwl =
+      std::dynamic_pointer_cast<SpatialDiscretization_FE>(discretization);
```
Since we are not casting to `SpatialDiscretization_PWL`, can we call this variable `fe` instead of `pwl`.

zachhardy replied:
Technically at the moment everything is `pwl` since only linear elements are supported, right (I.e. `pwld` and `pwlc` fall under `pwl`)? For anticipating future support for higher order elections, I agree with this change though.

Naktakala commented on ChiModules/LinearBoltzmannSolver/lbs_compute_precursors.cc:
```diff
@@ -0,0 +1,61 @@
+  precursor_new_local.assign(precursor_new_local.size(), 0.0);
+
+  //================================================== Loop over cells
+  for (auto& cell : grid->local_cells)
```
Might be silly, but can you make this `const auto&`, only if it does not conflict with something.

zachhardy replied:
Done.

Naktakala commented on ChiModules/LinearBoltzmannSolver/lbs_compute_precursors.cc:
```diff
@@ -0,0 +1,61 @@
+      exit(EXIT_FAILURE);
+    }
+
+    auto xs = material_xs[xs_id];
```
Please change this to a const reference: `const auto& xs = *material_xs[xs_id]` and change the related arrow operators.

zachhardy replied:
Done.

zachhardy committed:
Addressed reviewer comments.

1) Removed spurious C style cast.
2) pwl -> fe rename.
3) const reference to cross sections.

zachhardy committed:
Merge remote-tracking branch 'upstream/development' into cell-centered-precursors
